Spatial raster prep (#7)
ninoxconsulting/r-telemetry-workshop-nov-2023: 07-processing-vector-data.qmd

#   st_intersection(., aoi) %>% 
#       st_cast("MULTILINESTRING") 
#     } else .} 
#   
# st_write(roads, "clean_data/roads.gpkg")

```

We can assume that the type of road will have a different impact on Caribou avoidance or predator use. We can assign a value based on the road type. In this case we will assign a higher value to roads used more frequently.

```{r, eval = FALSE}

# read in roads
roads <- st_read(file.path("clean_data", "roads.gpkg"))

# assign a value to the roads based on estimated speed of travel or use

roads <- roads %>% 
  mutate(rd_value = case_when(
            ROAD_SURFACE == "loose" ~ 25,
            ROAD_SURFACE == "overgrown" ~ 5,
            ROAD_SURFACE == "rough" ~ 10,
            ROAD_SURFACE == "unknown" ~ 7.5)) 

```

Now we can convert this to a raster using the template we created already

```{r, eval = FALSE}
# convert roads to a raster 
rroads <- rasterize(roads, template, field = "rd_value" )

plot(rroads)

```
As the impact of the roads is likely to impact a wider influence than a single pixal we can use a moving window analysis to expand the influence of roads for a given value. 

```{r, eval = FALSE}
# create a moving window 

> rrdens <- focal(rroads, w=9, fun="sum", na.rm=TRUE)

Reviewer: Should maybe clarify exactly what using `w=9` means practically? Does that mean four pixels on either side of the road? At 25m pixels, adding 100m of "road influence" on either side of the road?

plot(rrdens)

terra::writeRaster(rrdens, file.path("clean_data", "road_density.tif"), overwrite = TRUE) 

```

## 2. Calculate distance from water

> While density provides a measure of impact we may also be interested in capturing the distance of fixes to caribou points. In this case we will explore the distance to water bodies within the study areas. 

Reviewer: ```suggestion
While density provides a measure of impact we may also be interested in capturing the distance of certain features to caribou points. In this case we will explore the distance to water bodies within the study areas. 
```

```{r, eval = FALSE}

# read in water 

> water <- st_read(file.path("clean_data", "roads.gpkg"))

Reviewer: Reading in roads here again

> # calculate the distance to water for each pixal in the raster

Reviewer: ```suggestion
# calculate the distance to water for each pixel in the raster
```

    dplyr::select(conif )

#plot(vri_conif$BCLCS_LEVEL_4)

st_write(vri_conif, file.path("clean_data","vri_conif.gpkg") , append = FALSE)

# Age class greater than 40 years
vri_ageclass <- vri %>%
    dplyr::mutate(age_class = as.numeric(PROJ_AGE_CLASS_CD_1)) %>%
    dplyr::filter(age_class >= 3) %>%
    dplyr::select(age_class)

st_write(vri_ageclass, file.path("clean_data", "vri_ageclass.gpkg"), append = FALSE)
 

# Crown closure class 
vri_cc <- vri %>%
    dplyr::mutate(cc_class = as.numeric(CROWN_CLOSURE_CLASS_CD)) %>%
    dplyr::select(cc_class)
st_write(vri_cc, file.path("clean_data", "vri_cc.gpkg"), append = FALSE)
 

```

```{r, eval = FALSE}

bec <- st_read(file.path("clean_data","bec.gpkg"))
vri_cc <- st_read(file.path("clean_data","vri_cc.gpkg"))
vri_conif <- st_read(file.path("clean_data","vri_conif.gpkg"))
vri_ageclass <- st_read(file.path("clean_data","vri_ageclass.gpkg"))
#water
#roads 
#cutblocks

# convert to rasters
vri_cc <- rasterize(vri_cc, template, field = "cc_class" )
vri_conif <- rasterize(vri_conif, template, field = "conif" )
vri_ageclass <- rasterize(vri_ageclass, template, field = "age_class" )

# stack into a set of rasters

> vect_stack <- c(vri_cc, vri_conif, vri_ageclass, water_dis, rrdens)

Reviewer: I think referring to this as `vect_stack` might be confusing, as it's not really a stack of vectors right? I know that you mean they're directly converted from vectors (and not calculate like the slope, aspect etc), but I wonder about a different word than vector to refer to them (same on line 218-220)

```

## Lets combine our already created rasters 

```{r, eval = FALSE}

> rslope <- rast(file.path("clean_data","slope.gpkg"))
> aspect <- rast(file.path("clean_data","aspect.gpkg"))
> tri  <- rast(file.path("clean_data","tri.gpkg"))
> trim_3005 <- rast(file.path("clean_data","dem.gpkg"))

Reviewer: These are tifs right?

```suggestion
rslope <- rast(file.path("clean_data","slope.tif"))
aspect <- rast(file.path("clean_data","aspect.tif"))
tri  <- rast(file.path("clean_data","tri.tif"))
trim_3005 <- rast(file.path("clean_data","dem.tif"))
```

Author: I was reading in the vector files you pull down from bc data cat so I think they should be read in as .gpk and exported as tif, so leaving it as is.

Reviewer: slope, aspect, dem, and tri were never vector though?

> # create a raster stack

Reviewer: This might be a foreign concept so probably worth spending some time explaining it

rast_stack <- c(trim_3005, rslope, aspect, tri)

```

## 4. Combine raster layers into a stack 

```{r, eval = FALSE}

> # lets combine both the raster stack and the vector stack 

Reviewer: See comment above. When I saw this I thought "Wow, you can stack vectors and rasters??" But I don't think that's right...

all_stack <- c(vect_stack, rast_stack)

plot(all_stack)

# we can write this out as a tif (raster object) 
#writeRaster(all_stack, file.path("clean_data", "rstack.tif"), overwrite = T)

> ## or we can write this out to a very small R object 

Reviewer: Probably worth touching on the pros and cons of the two approaches (size and ease of reading into R vs portability with other software)
